llvm/test/Transforms/VectorCombine/AArch64/load-extractelement-scalarization.ll

```diff
@@ -688,9 +688,9 @@ define i32 @load_multiple_extracts_with_constant_idx(ptr %x) {
 define i32 @load_multiple_extracts_with_constant_idx_profitable(ptr %x) {
 ; CHECK-LABEL: @load_multiple_extracts_with_constant_idx_profitable(
 ; CHECK-NEXT:    [[LV:%.*]] = load <8 x i32>, ptr [[X:%.*]], align 16
-; CHECK-NEXT:    [[E_0:%.*]] = extractelement <8 x i32> [[LV]], i32 0
-; CHECK-NEXT:    [[E_1:%.*]] = extractelement <8 x i32> [[LV]], i32 6
-; CHECK-NEXT:    [[RES:%.*]] = add i32 [[E_0]], [[E_1]]
+; CHECK-NEXT:    [[SHIFT:%.*]] = shufflevector <8 x i32> [[LV]], <8 x i32> poison, <8 x i32> <i32 6, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison, i32 poison>
```

Comment:
@davemgreen The comment above talks about scalarizing to avoid 2 v4i32 loads - but the cost model still does this - which is more realistic?

Reply:
Hello. Am I right in saying that the load is not included in the cost? It will be difficult to beat scalarization of the load if this doesn't lead to other optimizations, but I suspect that is not really what the costs are measuring.

Maybe @fhahn remembers more about this specific case. The change you have (pass the mask to the shuffle cost) seems like a sensible optimization. There is a comment that says:
```
  // Aggressively form a vector op if the cost is equal because the transform
  // may enable further optimization.
  // Codegen can reverse this transform (scalarize) if it was not profitable.
```
Maybe it should be more aggressive in the backend at scalarizing. It looks like the costs here should be extract-lane-0 + extract-lane-2 (2) + i32 add (1)   vs   extract-lane-0 + shuffle (1 now?) + v8i32 add (2).  If it could realize that the last v8i32 add was actually a v4i32 add, that might be more accurate (if I have those costs correct).

Reply:
In this case it can be a v2i32 add - and maybe the shuffle will fold away entirely as a 64-bit vector subvector extract - not sure, its a long time since I worked with (aarch32) neon ops!

Reply:
I don't believe there is a way to specify demanded elts to the cost functions, but would it be possible to manually specify the expected length of the add (maybe using the known register vector length if the length of the vector is longer)? It doesn't help with the cost/scalarization of the load, but might make the cost overall a little more accurate.

Reply:
There's no demanded elts - but the shuffle mask is now all -1 apart from the single insertion index. It won't help for anything but the getShuffleCost of course.

Reply:
Yeah the extract might be cost=2 and the shuffle cost=1 (because of the perfect shuffle tables), which wont be helping either.

Reply:
I've updated the getShuffleCost call to include the shuffle's source vector value - so the costmodel can try to recognise loads.

Reply:
What about the cost of VectorOpCost, if it will be a smaller vector that gets used for the binop? Will lowering that prevent this triggering elsewhere?

And do you know how this transform "may enable further optimization"? Maybe that would make this worth while anyway.

Reply:
AFAICT "may enable further optimization" mainly means once we've folded to a shuffle, vectorcombine can do more with it.

Reply:
I've updated the AArch64 cost model to assume extractelement of a load is free if all users are extractelements - is that satisfactory do you think?

Reply:
I can pull out that costmodel change into its own PR if necessary.

Reply:
@davemgreen any thoughts on this - is the aarch64 costmodel change the way to go?

Reply:
Hello. I think it should be "if the vector load after legalization has a single extract - then it will be scalarized". Multiple extracts still don't scalarize in general. I'm not sure if this is the best way forward or not though. There are limits to what the cost model can handle well, and this patch (minus the aarch64 cost model change) looks like a step in the right direction.

Am I correct that shrinking the type used to cost the vector binop, to express that the whole (possibly multi-)vector width is not needed only a single lane, does not work well in practice?

If so maybe we go with the original. I can try another test out too, to see if single element shuffles should be costed more like lane moves.

Reply:
I've dropped the CostModel change.

Changing types is tricky - I've always kept to using types already existing somewhere in the pattern (or very occasionally int<->fp equivalents) and not make a guess at a new 'optimal' types. The best we could do is split them if the upper half is known undef/undemanded, but I'm still unsure how well it'd work even with a cost model to guide it.

Reply:
Yeah - I was thinking of getNumberOfParts / getRegisterBitWidth like the SLP vectorizer uses, and clamping the size of the vector type to one register.